target_snowflake/__init__.py

```diff
@@ -299,6 +316,13 @@ def flush_streams(
     for stream in streams_to_flush.keys():
         streams[stream] = {}
 
+        # update flushed_state position if we have state information for the stream
+        if stream in state:
```

Comment:
@koszti in case of postgres source, the state value has `bookmarks` and `currently_syncing` keys, example:
`{"bookmarks": {"postgres_source_db-logical1-logical1_table1": {"last_replication_method": "LOG_BASED", "lsn": 25346960, "version": 1569418760524, "xmin": null}, "postgres_source_db-logical1-logical1_table2": {"last_replication_method": "LOG_BASED", "lsn": 25347344, "version": 1569418760548, "xmin": null}, "postgres_source_db-logical2-logical2_table1": {"last_replication_method": "LOG_BASED", "lsn": 1, "version": 1569418760563, "xmin": null}}, "currently_syncing": null}
`

So this if statement would always be evaluated as False because the streams are in the bookmarks not in state. This also raises another question, in singer specs:

> STATE Message

> STATE messages contain the state that the Tap wishes to persist. STATE messages have the following properties:

> * value Required. The JSON formatted state value

> The semantics of a STATE value are not part of the specification, and should be determined independently by each Tap.

The last sentence means that we cannot have one way of checking if stream is in state because the value structure depends on the tap.

wdyt?

Reply:
Good catch, thanks for this. I've checked tap-postgres, tap-mysql, tap-snowflake, tap-zendesk, tap-jira, tap-oracle, and they all use the above `{"bookmarks": {..}, "currently_syncing": ..}` format.

Furthermore luckily they all use the common `write_bookmark` function from the common base singer python library at  https://github.com/singer-io/singer-python/blob/master/singer/bookmarks.py#L10

If the spec doesn't define the state value structure but the current taps are in sync, I think we can take this structure as the PipelineWise default and we need to check the new structure when we add new tap connectors. PipelineWise is still a bunch of pre-selected, forked and customised tap and target connectors.

```
if stream in stage.get('bookmarks', {}):
    flushed_state['bookmarks'][stream] = state[stream]
```
Or maybe there is a more elegant function in the singer library to update state messages:
 * https://github.com/singer-io/singer-python/blob/master/singer/bookmarks.py
 * https://github.com/singer-io/singer-python/blob/master/singer/messages.py
```

Reply:
Thank you for checking the other taps, it's great that they all have the same structure.
I think we can agree to have `{"bookmarks": {..}, "currently_syncing": ..}` format as Pipelinewise default and only supported format and check that every new tap type respects this.

I think we can get inspired by this singer function:

    def reset_stream(state, tap_stream_id):
        state = ensure_bookmark_path(state, ['bookmarks', tap_stream_id])
        state['bookmarks'][tap_stream_id] = {}
        return state

and add this function to update a stream:

    def update_stream(state, tap_stream_id, new_value):
        state = ensure_bookmark_path(state, ['bookmarks', tap_stream_id])
        state['bookmarks'][tap_stream_id] = new_value
        return state


The we can use it as follows:

    if stream in state['bookmarks']:
         flushed_state = update_stream(flushed_state, stream, state['bookmarks'][stream])

Reply:
thanks, I'll update in the PR. I'll try to add some more test cases too with state messages in the test streams.